### ISSUE TITLE
mallocMC: More Variables

### ISSUE THREAD

**Issue by ax3l** [ComputationalRadiationPhysics/cmake-modules#3]
mallocMC versions after 2.1 (or later) will probably compile [halloc](https://github.com/ComputationalRadiationPhysics/mallocMC/issues/81) as a library.

This module should then allow to compile every time (`thirdParty/` directory in PIConGPU) or to set as usual the (still missing) CMake variables:
- `mallocMC_DEFINITIONS`
- `mallocMC_LIBRARIES`

**Comment by ax3l** [comment]
@slizzered in case you have time for that in the next two weeks, feel free to catch this issue and to provide a pull request :)

**Comment by slizzered** [comment]
I will find some time after the 26th of March, so if you can wait that long...

**Comment by ax3l** [comment]
sure that is fine, thank you! :+1:

**Comment by slizzered** [comment]
I've done some additions to the file, but I don't succeed in building PIConGPU with mallocMC+halloc. Halloc requires to [use "separable compilation" for device code](https://github.com/canonizer/halloc/blob/master/README#L64-L72). For me, that usually works with setting `set(CUDA_SEPARABLE_COMPILATION ON)`. However, if I try to add it in PIConGPU's `CMakeLists.txt`, the configuration step already fails. This might be a bug in CMake or the shipped `FindCUDA.cmake` file, but I'll have to investigate further.

**Comment by ax3l** [comment]
the "separable compilation" is not an option that you could forward through the cmake module anyway.

one thing could be, that mallocMC needs to be compiled in advance - and not as a "header only `thirdParty` module" within PIConGPU (or still within `thirdParty` but with a "separable compilation" in advance for halloc) and you simply provide additional flags for
- `mallocMC_DEFINITIONS`
- `mallocMC_LIBRARIES`

via cmake-modules. is that possible?

I am not sure if they mean `CUDA_SEPARABLE_COMPILATION` when they say you have to compile halloc "separately". at least the example you posted does suggest my case.

**Comment by slizzered** [comment]
I agree that halloc needs to be compiled in advance, that was already planned. Same goes for the 2 variables:
- `mallocMC_DEFINITIONS`: `-lhalloc` `-L/.../[Path to libhalloc.a]`
- `mallocMC_LIBRARIES:` `[Path to folder where libhalloc.a is located.]`

The problem arises when trying to link PIConGPU itself after all compilations are completed:

```
/home/carli/dev/picongpu/src/picongpu/../../thirdParty/mallocMC/halloc/bin/libhalloc.a(tmpxft_00006be0_00000000-44_halloc.o): In function `__sti____cudaRegisterAll_53_tmpxft_00006be0_00000000_24_halloc_compute_35_cpp1_ii_fda90288()':
tmpxft_00006be0_00000000-10_halloc.compute_20.cudafe1.cpp:(.text.startup+0x15): undefined reference to `__cudaRegisterLinkedBinary_53_tmpxft_00006be0_00000000_24_halloc_compute_35_cpp1_ii_fda90288'
```

halloc uses some features that require relocatable device code linking, in addition to compiling. Compilation of a project that uses halloc (and by extension mallocMC with halloc-configuration) needs to do the final linking of the executable in a separate step (similar to [1] and [2], although for other reasons).

CMake can hide/abstract this 2-step approach when you set the "separate compilation" option in the project's (PIConGPU's) `CMakeLists.txt`. But the flag is somehow incompatible with some other configuration in our case. It did work for other projects.

The problem does not arise as long as 
 **a)** we don't use halloc in PIConGPU or
 **b)** all projects that want to use halloc find a way to compile with the SEPARABLE_COMPILATION flag (I did some test setups that compiled and linked successfully)

Not a really big issue now, but it might become a problem in the future...

[1] http://stackoverflow.com/questions/22115197/dynamic-parallelism-undefined-reference-to-cudaregisterlinkedbinary-linking
[2] http://stackoverflow.com/questions/22076052/cuda-dynamic-parallelism-makefile

**Comment by ax3l** [comment]
Interesting, thank you for the links!
Just for my curiosity: does halloc use dynamic parallelism?

A hint what the cmake macros stand for, since your examples are wrong. it should be:
- `mallocMC_DEFINITIONS`: cmake list of `-DSOMEDEFINES=STH;-DOTHERDEF`
- `mallocMC_LIBRARIES`: cmake list of `/absolute/path/to/libhalloc.a;/path/to/other/libOther.so`

examples for defines can be found in [FindPNGwriter.cmake](https://github.com/ComputationalRadiationPhysics/cmake-modules/blob/232cad5508e8d669ee2b59546832f3040b978ead/FindPNGwriter.cmake#L126) or [FindSplash.cmake](https://github.com/ComputationalRadiationPhysics/cmake-modules/blob/232cad5508e8d669ee2b59546832f3040b978ead/FindSplash.cmake#L118) and for libraries in [FindADIOS.cmake](https://github.com/ComputationalRadiationPhysics/cmake-modules/blob/232cad5508e8d669ee2b59546832f3040b978ead/FindADIOS.cmake#L125-L160)

To trigger the `SEPARABLE_COMPILATION` one can define a `mallocMC_HAS_HALLOC` or `mallocMC_NEEDS_SEPARABLE_COMPILATION` variable in the cmake module of `mallocMC` that the user can then check in PIConGPU's `CMakeLists.txt` to enable `SEPARABLE_COMPILATION` (besides the fact that it does not compile currently, which is an other issue).

**Comment by slizzered** [comment]
FYI: halloc (at least the library itself) does not use dynamic parallelism (allocation is done through `__device__`  functions)

Thanks for the hint, I did some reading and set it correctly now (I hope).

What was also done:
- I reworked the module structure to make it more stable and with less user-defined variables and messages
- REQUIRED and QUIET works as intended
- halloc is now available as an optional COMPONENT
  - if halloc is found, the module automatically executes `set(SEPARABLE_COMPILATION ON PARENT_SCOPE)`
  - halloc can be non-required even if mallocMC itself is required. Just split it into 2 lines:

``` cmake
find_package(mallocMC 2.1.0 REQUIRED)
find_package(mallocMC 2.1.0 COMPONENTS halloc)
```

Note, that the code above will activate separable compilation, if halloc was found.

**Comment by ax3l** [comment]
oh that is cool, I did not know

  set(varname ON PARENT_SCOPE)

exists.
